RuminantFarmSystems/RuFaS
# Suppress log files and specify log directory for tasks and the Task Manager

This PR adds two features to more effectively handle outputs: a switch to prevent logs from being written to output files, and the ability to specify where logs from the Task Manager are written to.

Note that I will update the wiki to reflect the changes that have been made in this PR, for both the GNU args and the change to the task metadata properties.

### Context

Issue(s) closed by this pull request: closes #1397 

      
How big are the changes in the PR? Would you nominate it for a major version upgrade?

### What

Users will be able to specify if they want the logs generated by individual tasks and the Task Manager to be written to log files. Logs in this case means all the logs added to the Output Manager, and also the metadata used, variable names, and `get_data` usage counts.

Users can also specify where the Task Manager writes log files to - making its behavior and capabilities more similar to those of individual tasks.

### Why

As discussed in #1397, it is handy to be able to sort and limit which output files are written when running a simulation (or multiple simulations).

### How

To specify whether or not to write logs to a file, users put a bool argument in a task called "suppress_log_files" which defaults to False. That will determine whether or not logs are written to files for that individual task. To stop the Task Manager from writing its logs to files, a new GNU argument has been added.

Another GNU argument has been added which allows users to specify the path to the directory they want the Task Manager's logs to be written to. The behavior of this option is very similar to the way that the log directory is set for individual tasks.

### Test plan

The test suite has been updated and rerun. Also, I have:
- Raised exceptions in both the Task Manager and in the default simulation. I checked that logs are automatically dumped into files, overriding the user's specification of whether or not log files should be written.
- Run the default simulation with "suppress_log_files" turned on and off to verify that logs are not written and written to files, respectively.
- Run the default simulation with the "-s" GNU args specified and unspecified to verify that logs from the Task Manager are not written and written to files, respectively.
- Specified a directory in the GNU args for Task Manager's logs to be written to, and checked that the logs were in fact written there.
